### ISSUE TITLE
Multi tenancy return other's data

### ISSUE THREAD

**Issue by wenerme** [cube-js/cube.js#4129]
**Describe the bug**

setup cube multi tenancy like below, but seem's cube will return data for other datasource in first few request.

event, throw error in driverFactory, cube still return data for last request datasource.

1. request tenant A - return tenant A data
2. request tenant B - return tenant A data
3. request tenant B - return tenant B data - after several request
4. request tenant C - throw error in driverFactory - **still** return tenant B data
    - log error Error while renew cycle

**options**

```ts
const ops = {
  webSockets: true,

  cacheAndQueueDriver: 'memory',
  devServer: isDev,

  scheduledRefreshTimeZones: ['Asia/Shanghai'],
  scheduledRefreshContexts: async () => {
    // loop all known tenants
    const names = Object.entries(tenants)
      .filter(([k, v]) => !v.disabled)
      .map(([k, v]) => k);
    return names.map((v) => ({
      securityContext: {
        tenantId: v,
        dataSource: v,
      },
    }));
  },

    contextToAppId: (ctx) => {
      let tenantName = getTenantName(ctx);
      if (!tenantName) {
        throw new Error(`invalid request: no tenant`);
      }
      if (!ctx.securityContext && !isDev) {
        throw new Error(`unauthenticated`);
      }
      ctx.securityContext ??= {};
      ctx.securityContext['tenantId'] = tenantName;
      ctx.securityContext['dataSource'] = tenantName;
      
      let appId = `CUBEJS_APP_${tenantName}_${ctx.securityContext['preferred_username'] ?? ctx.securityContext['sub']}`;
      console.log(`context to tenant`, tenantName, appId);

      return appId;
    },

  dbType: (ctx) => 'postgres',
  driverFactory: (ctx) => {
    const ds = dataSources[ctx.dataSource]
    if (!ds) {
      throw new Error('datasource not found');
    }
    return new PostgresDriver(ds);
  },

  externalDbType: 'postgres',
  externalDriverFactory: () => {
    return new PostgresDriver({});
  },
  externalStore: {
    type: 'postgres',
  },
  jwt: {
    algorithms: ['RS256', 'HS256'],
    jwkUrl: (payload) => {
      return getJwkURLOfIssuer(payload.payload?.iss)
    },
  },
}
```

**cube**

```js
const {
  securityContext: { tenantId, dataSource = 'default' },
} = COMPILE_CONTEXT;

console.info(`compile schema`, tenantId, dataSource);

export const BaseModel = cube({
  dataSource,
});

cube(`Users`, {
  sql: `SELECT * FROM users`,
  sqlAlias: 'users',
  extends: BaseModel,
});
```

**To Reproduce**

**Expected behavior**

**Screenshots**

**Version:**

v0.29.28

**Additional context**

**Comment by ivan-vdovin** [comment]
Hi @wenerme !
What getTenantName(ctx) looks like here?

**Comment by wenerme** [comment]
@ivan-vdovin 

something like this

```ts
function getTenantName(ctx: RequestContext) {
  if (!ctx.securityContext) {
    throw new Error('unauthenticated');
  }
  if (!ctx.securityContext.iss) {
    throw new Error('invalid auth');
  }
  return new URL(ctx.securityContext.iss).host.replace(/[.]/g, '_');
}
```

the tenant is detected from issuer host.

**Comment by paveltiunov** [comment]
@wenerme Queries you use would be really helpful here.

**Comment by paveltiunov** [comment]
@wenerme https://cube.dev/docs/config#context-to-orchestrator-id is missing in your configuration.

**Comment by hassankhan** [comment]
Thanks for reporting @wenerme 👍 Fixed in #4760
